### PR TITLE
Listening for toasts via `OnAccessibilityEventListener`

### PR DESCRIPTION
**Pull request by vmaxim**
Reimplement notifications listener through `UiAutomation.OnAccessibilityEventListener`.

Current implementation is slightly flaky and sometimes toast messages could not be catched even on real device. Also```onAccessibilityEventListener``` eliminates the events sync problems like a appium/appium#9328.

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -93,12 +93,12 @@ dependencies {
 }
 
 task installAUT(type: Exec) {
-    group = 'test'
-    description = 'Install app under test'
+    group = "test"
```
Comment (mykola-mokhnach) [inline app/build.gradle]: single quotes are ok. One only uses double quotes in Groovy together with templates

File: app/src/main/java/io/appium/uiautomator2/core/UiAutomation.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+    private UiAutomation() {
+        uiAutomation = UiAutomatorBridge.getInstance().getUiAutomation();
+    }
+
+    public static UiAutomation getInstance() {
+        return INSTANCE;
+    }
+
+    public OnAccessibilityEventListener getOnAccessibilityEventListener() {
```
Comment (mykola-mokhnach) [inline app/src/main/java/io/appium/uiautomator2/core/UiAutomation.java]: Cannot we just put it inside UiAutomatorBridge?

File: app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java [hunk 3]
```diff
@@ -1,114 +1,108 @@
...
+    private static final int TOAST_CLEAR_TIMEOUT = 3500;
 
-public final class NotificationListener {
-    private final static NotificationListener INSTANCE = new NotificationListener();
-    private static List<CharSequence> toastMessages = new ArrayList<CharSequence>();
-    private final int TOAST_CLEAR_TIMEOUT = 3500;
-    private final int WAIT_FOR_EVENT_TIMEOUT = 500;
-    public boolean isListening = false;
-    private Listener listener;
+    private List<CharSequence> toastMessage = null;
```
Comment (mykola-mokhnach) [inline app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java]: I'd rather assign it to an empty list.

File: app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java [hunk 4]
```diff
@@ -1,114 +1,108 @@
...
+    private long previousTime = currentTimeMillis();
+    private OnAccessibilityEventListener originalListener = null;
+    private UiAutomation uiAutomation;
```
Comment (mykola-mokhnach) [inline app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java]: final?

File: app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java [hunk 5]
```diff
@@ -1,114 +1,108 @@
...
 
-public final class NotificationListener {
-    private final static NotificationListener INSTANCE = new NotificationListener();
-    private static List<CharSequence> toastMessages = new ArrayList<CharSequence>();
-    private final int TOAST_CLEAR_TIMEOUT = 3500;
-    private final int WAIT_FOR_EVENT_TIMEOUT = 500;
-    public boolean isListening = false;
-    private Listener listener;
+    private List<CharSequence> toastMessage = null;
+    private long previousTime = currentTimeMillis();
```
Comment (mykola-mokhnach) [inline app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java]: what is the purpose of this property?

File: app/src/main/java/io/appium/uiautomator2/core/UiAutomation.java (hunk 2 above)
Comment (vmaxim) [reply app/src/main/java/io/appium/uiautomator2/core/UiAutomation.java]: Assigned listener can't be retrieved from UiautomatorBridge.

File: app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java (hunk 5 above)
Comment (vmaxim) [reply app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java]: it stores the time when the last toast was caught and used to manage toast's lifetime.

Comment (mykola-mokhnach) [reply app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java]: then I'd rather rename it to something like recentToastTimestamp

File: app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java [hunk 6]
```diff
@@ -1,114 +1,108 @@
...
-                    toastMessages = accessibilityEvent.getText();
-                    previousTime = currentTimeMillis();
-                }
-                if (stopLooping) {
-                    break;
-                }
+    @Override
+    public synchronized void onAccessibilityEvent(AccessibilityEvent event) {
+        if (event.getEventType() == AccessibilityEvent.TYPE_NOTIFICATION_STATE_CHANGED) {
+            Logger.debug("Catch toast message:" + event);
```
Comment (jlipps) [inline app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java]: nit: probably want space after `:`

File: app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java [hunk 7]
```diff
@@ -1,114 +1,110 @@
...
+    }
 
-        public void stopLooping() {
-            stopLooping = true;
+    public boolean isListening() {
+        return uiAutomation.getOnAccessibilityEventListener() == this;
+    }
+
+    protected long getToastClearTimeout() {
+        return TOAST_CLEAR_TIMEOUT;
```
Comment (mykola-mokhnach) [inline app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java]: is this timeout always hardcoded or we can get its actual value from the system?

File: app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java [hunk 8]
```diff
@@ -1,114 +1,110 @@
...
+    }
+
+    @NonNull
+    protected List<CharSequence> getToastMessage() {
+        if (!toastMessage.isEmpty() && currentTimeMillis() - recentToastTimestamp > getToastClearTimeout()) {
```
Comment (mykola-mokhnach) [inline app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java]: is it really necessary to delete the message after the timeout? I'd rather collect toasts together with their timestamps and let the client to decide whether he needs them or not by analyzing the timestamp

File: app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java (hunk 7 above)
Comment (vmaxim) [reply app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java]: There is no direct way to obtain this timeout from the system.

File: update-gradle-version.js [hunk 9]
```diff
@@ -23,9 +23,9 @@ if (!/\d+\.\d+\.\d+/.test(version)) {
 console.log(`Updating gradle build file to version '${version}'`);
 replace({
   files: GRADLE_BUILD_FILE,
-  from: /^\s+versionName\s"(.+)"$/gm,
+  from: /^\s+versionName\s'(.+)'$/gm,
```
[the remark this answers is not in the document]
Comment (vmaxim) [reply update-gradle-version.js]: https://github.com/appium/appium-uiautomator2-server/pull/158/files#diff-39e7d8c00954e920b98e7636f0ac30b2R29

File: app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java (hunk 8 above)
Comment (vmaxim) [reply app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java]: I'll be happy if someone implements this behaviour :)

Comment (mykola-mokhnach) [reply app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java]: I believe you can easily handle it on your own ;)

File: app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java [hunk 10]
```diff
@@ -1,114 +1,110 @@
...
-    private final int WAIT_FOR_EVENT_TIMEOUT = 500;
-    public boolean isListening = false;
-    private Listener listener;
+    private List<CharSequence> toastMessage = new CopyOnWriteArrayList<>();
+    private long recentToastTimestamp = currentTimeMillis();
+    private OnAccessibilityEventListener originalListener = null;
+    private final UiAutomation uiAutomation;
 
-    private NotificationListener() {
+    NotificationListener() {
```
Comment (mykola-mokhnach) [inline app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java]: why not keeping it private?

Comment (mykola-mokhnach) [reply app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java]: also codacy does not appreciate package-level modifiers

Comment (vmaxim) [reply app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java]: It has package private modifier for testing purposes. Also there is inconsistency between android lint and codacy. You can check it out from android studio (```Analyze > Inspect Code```) or command line (```./gradlew lintDebug```).

**Comment by vmaxim** [comment]
Travis has gone crazy.

**Comment by imurchie** [comment]
Only for this repo. Weird.

File: app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java (hunk 10 above)
Comment (mykola-mokhnach) [reply app/src/main/java/io/appium/uiautomator2/model/NotificationListener.java]: I know, but I'd rather stick to Codacy

**Pull request merged**